### PR TITLE
C++: Add cases in the Allocation model.

### PR DESCRIPTION
**Pull request by geoffw0**
While working on the https://github.com/github/codeql/pull/4783 I noticed a few holes in our allocation models.  I've also added test cases for `strdup`, a standard library function which it turns out we already support.

### CODE REVIEW

File: cpp/ql/src/semmle/code/cpp/models/implementations/Allocation.qll [hunk 1]
```diff
@@ -82,7 +82,9 @@ private class AllocaAllocationFunction extends AllocationFunction {
     hasGlobalName([
         // --- stack allocation
         "alloca", // // alloca(size)
-        "__builtin_alloca" // __builtin_alloca(size)
+        "__builtin_alloca", // __builtin_alloca(size)
+        "_alloca", // __alloca(size)
```
Comment (MathiasVP) [inline cpp/ql/src/semmle/code/cpp/models/implementations/Allocation.qll:86]: Is it a mistake that the `"_alloca"` string with 1 underscore when the comment has 2?

**Review by MathiasVP — COMMENTED** [review]
Looks good! I only have one comment (and then we need an internal PR as well, of course).

File: cpp/ql/src/semmle/code/cpp/models/implementations/Allocation.qll (hunk 1 above)
Comment (geoffw0) [reply cpp/ql/src/semmle/code/cpp/models/implementations/Allocation.qll]: Good spot.  One underscore is correct (see https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/alloca?view=msvc-160), I have corrected the comment.

**Comment by geoffw0** [comment]
There was a failing (internal) test and I had to remove support for `SysAllocString`. It turns out supporting all of the SysAlloc stuff pushed beyond the flexibility offered by our current models, supporting just `SysAllocString` and `SysFreeString` leads to problems where the code uses other functions as well, but supporting just `SysFreeString` as we previously did is safe (and it is rarely used to free stuff that was allocated in ways we do model, I think).

When we have support for access paths in our models, I think we should be able to do better.

**Review by jbj — APPROVED** [review]

**Pull request merged**